ui/src/shared/components/AnnotationFilterControlInput.tsx

```diff
@@ -0,0 +1,157 @@
+import React, {PureComponent, ChangeEvent, KeyboardEvent} from 'react'
+
+import {ClickOutside} from 'src/shared/components/ClickOutside'
+
+interface Props {
+  value: string
+  suggestions: string[]
+  onChange: (newValue: string) => void
+  onSelect: (newValue: string) => void
+  onFocus: () => void
+  inputClass?: string
+}
+
+interface State {
+  selectionIndex: number | null
+  areSuggestionsVisible: boolean
+  filteredSuggestions: string[]
+}
+
+const NUM_SUGGESTIONS = 8
+
+class AnnotationFilterControlInput extends PureComponent<Props, State> {
+  public static getDerivedStateFromProps(props) {
+    const {suggestions, value} = props
+
+    const filtered: string[] = suggestions.filter(
+      (v, i) => v.includes(value) && i < NUM_SUGGESTIONS
+    )
+    const unique: string[] = [...new Set(filtered)]
+
+    unique.sort((a, b) => a.localeCompare(b))
+
+    return {filteredSuggestions: unique}
+  }
+
+  constructor(props) {
+    super(props)
+
+    this.state = {
+      filteredSuggestions: props.suggestion,
+      selectionIndex: 0,
+      areSuggestionsVisible: false,
+    }
+  }
+
+  public render() {
+    const {value} = this.props
+
+    return (
+      <ClickOutside onClickOutside={this.handleClickOutside}>
+        <div className="suggestion-input">
+          <input
+            className={this.inputClass}
+            value={value}
+            onChange={this.handleChange}
+            onFocus={this.handleFocus}
+            onKeyUp={this.handleKeyUp}
+          />
+          <div
+            className={`suggestion-input--suggestions dropdown-menu ${
+              this.suggestionsClass
+            }`}
+          >
+            {this.renderSuggestions()}
+          </div>
+        </div>
+      </ClickOutside>
+    )
+  }
+
+  private renderSuggestions() {
+    const {onSelect} = this.props
+    const {selectionIndex, filteredSuggestions} = this.state
+
+    return filteredSuggestions.map((v, i) => {
```

Comment:
what do you think about making an intermediate var for this? it's a bulky `return`